src/kimmdy/runmanager.py

```diff
@@ -772,21 +815,57 @@ def _apply_recipe(self, files: TaskFiles) -> TaskFiles:
             write_time_marker(files.outputdir / MARK_FAILED, "failed")
             raise RuntimeError(m)
 
+        if isinstance(self.kmcresult, KMCRejection):
+            m = "Attampting to apply a recipe where none has been accepted (KMCRejection)."
+            logger.error(m)
+            raise RuntimeError(m)
+
         recipe = self.kmcresult.recipe
         logger.info(f"Start Recipe in KIMMDY iteration {self.iteration}")
         logger.info(f"Recipe: {recipe.get_recipe_name()}")
         logger.debug(f"Performing recipe steps:\n{pformat(recipe.recipe_steps)}")
 
         # Set time to chosen 'time_start' of KMCResult
         ttime = self.kmcresult.time_start
+        plugin_time_index = self.kmcresult.time_start_index_within_plugin
```

Comment:
same here. 200 lines for something called _apply_recipe seems excessive. If we really need to refactor, that would be something for another day, I am just wondering whether the code is in the right place

Reply:
Absolutely agree, I'm also not a fan of the coupling we have going on between the tasks right now. E.g. applying the recipe depends on the state created by choosing the recipe, which relies on asking the reaction plugins. So those should be a single task, which is only broken down for readability into smaller functions. But it shouldn't be possible to accidentally try to apply recipes without having chosen them. On the other hand it would be still be nice to be able to call those sub-tasks individually for testing and interactive exploration.

For example, our runmanager iterator allows us to step through one by one like this: https://github.com/graeter-group/kimmdy-examples/blob/2b2b8fd5b1c579a0dc17df6a5ca39e8feec47599/wip.qmd#L174-L250

We should use the same approach to write tests for complete kimmdy runs that also test assertions during after each task instead of only checking the log at the end.

Reply:
I created issues for those